chain/client/src/client.rs

```diff
@@ -444,77 +449,57 @@ impl Client {
         Ok(())
     }
 
-    /// Check that this block height is not known yet.
-    fn known_block_height(&self, next_height: BlockHeight, known_height: BlockHeight) -> bool {
-        #[cfg(feature = "test_features")]
-        {
-            if self.adv_produce_blocks {
-                return false;
-            }
-        }
-
-        next_height <= known_height
-    }
-
-    /// Check that we are next block producer.
-    fn is_me_block_producer(
+    /// Checks couple conditions whether Client can produce new block on height
+    /// `height` on top of block with `prev_header`.
+    /// Needed to skip several checks in case of adversarial controls enabled.
+    /// TODO: consider returning `Result<(), Error>` as `Ok(false)` looks like
+    /// faulty logic.
+    fn can_produce_block(
         &self,
+        prev_header: &BlockHeader,
+        height: BlockHeight,
         account_id: &AccountId,
         next_block_proposer: &AccountId,
-    ) -> bool {
+    ) -> Result<bool, Error> {
         #[cfg(feature = "test_features")]
         {
-            if self.adv_produce_blocks_only_valid {
-                return account_id == next_block_proposer;
-            }
-            if self.adv_produce_blocks {
-                return true;
+            if self.adv_produce_blocks == Some(AdvProduceBlocksMode::All) {
+                return Ok(true);
             }
         }
 
-        account_id == next_block_proposer
-    }
-
-    fn should_reschedule_block(
-        &self,
-        prev_hash: &CryptoHash,
-        prev_prev_hash: &CryptoHash,
-        height: BlockHeight,
-        known_height: BlockHeight,
-        account_id: &AccountId,
-        next_block_proposer: &AccountId,
-    ) -> Result<bool, Error> {
-        if self.known_block_height(height, known_height) {
-            return Ok(true);
-        }
-
-        if !self.is_me_block_producer(account_id, next_block_proposer) {
+        // If we are not block proposer, skip block production.
+        if account_id != next_block_proposer {
             info!(target: "client", "Produce block: chain at {}, not block producer for next block.", height);
```

Comment:
nit: Use the structured log syntax by moving height to a dedicated field
info!(target: "client", height, "Skipping block production, not block producer for next block.");